atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/RowOrDynamicColumnRenderer.java

```diff
@@ -260,9 +266,29 @@ private void prefix(int i, boolean isSorted) {
         } line("}");
     }
 
-    private void renderComputeFirstComponentHash() {
-        NameComponentDescription firstRowPart = getRowPartsWithoutHash().get(0);
-        line("long ", NameMetadataDescription.HASH_ROW_COMPONENT_NAME, " = Hashing.murmur3_128().hashBytes(ValueType.", firstRowPart.getType().name(), ".convertFromJava(", varName(firstRowPart), ")).asLong();");
+    private void renderComputeFirstNComponentHash(int numberOfComponentsHashed) {
+        List<NameComponentDescription> components = getRowPartsWithoutHash().subList(0, numberOfComponentsHashed);
+        List<String> vars = Lists.newArrayList();
+        for (NameComponentDescription comp : components) {
+            vars.add(varName(comp));
+        }
+        line("long ", NameMetadataDescription.HASH_ROW_COMPONENT_NAME, " = computeHashFirstComponents(", Joiner.on(", ").join(vars), ");");
+    }
+
+    private void renderComputeFirstNComponentsHashMethod(int numberOfComponentsHashed) {
+        List<NameComponentDescription> components = getRowPartsWithoutHash().subList(0, numberOfComponentsHashed);
+        line("public static long computeHashFirstComponents"); renderParameterList(components); lineEnd(" {"); {
+            List<String> vars = Lists.newArrayList();
+            for (NameComponentDescription comp : components) {
+                String var = varName(comp) + "Bytes";
+                vars.add(var);
+                line("byte[] ", var, " = ", comp.getType().getPersistCode(varName(comp)), ";");
+                if (comp.getOrder() == ValueByteOrder.DESCENDING) {
+                    line("EncodingUtils.flipAllBitsInPlace(", var, ");");
```

Comment:
FlipAllBitsInPlace does the flipping in place :), but for some reason it also returns the same reference array. So re-assigning it would make no difference.